go/runtime/committee/client.go

```diff
@@ -0,0 +1,406 @@
+package committee
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/resolver"
+
+	"github.com/oasislabs/oasis-core/go/common/crypto/signature"
+	cmnGrpc "github.com/oasislabs/oasis-core/go/common/grpc"
+	"github.com/oasislabs/oasis-core/go/common/grpc/resolver/manual"
+	"github.com/oasislabs/oasis-core/go/common/identity"
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	"github.com/oasislabs/oasis-core/go/common/node"
+	"github.com/oasislabs/oasis-core/go/common/pubsub"
+)
+
+// NodeSelectionFeedback is feedback to the node selection policy.
+type NodeSelectionFeedback struct {
+	// ID is the node identifier.
+	ID signature.PublicKey
+
+	// Bad being non-nil signals that the currently selected node is bad and contains the reason
+	// that lead to the decision.
+	Bad error
+}
+
+// NodeSelectionPolicy is a node selection policy.
+type NodeSelectionPolicy interface {
+	// UpdateNodes updates the set of available nodes.
+	UpdateNodes([]signature.PublicKey)
+
+	// UpdatePolicy submits feedback to the policy which can cause the policy to update its current
+	// node selection.
+	UpdatePolicy(feedback NodeSelectionFeedback)
+
+	// Pick picks a node from the set of available nodes accoording to the policy.
+	Pick() signature.PublicKey
+}
+
+type roundRobinNodeSelectionPolicy struct {
+	sync.Mutex
+
+	nodes []signature.PublicKey
+	index int
+}
+
+func (rr *roundRobinNodeSelectionPolicy) UpdateNodes(nodes []signature.PublicKey) {
+	rr.Lock()
+	rr.nodes = nodes
+	rr.index = 0
```

Comment:
Won't this cause every client to hit the same node?

Reply:
Good idea, I added a random shuffle.